includes/class-wcsatt-product.php

```diff
@@ -0,0 +1,856 @@
+			self::set_product_property( $product, 'subscription_period_interval', null );
+			self::set_product_property( $product, 'subscription_length', null );
+
+			$scheme_set = true;
+		}
+
+		do_action( 'wcsatt_set_product_subscription_scheme', $key, $active_scheme_key, $product );
+
+		return $scheme_set;
+	}
+
+	/**
+	 * Set the product as purchasable on a recurring basis only.
+	 *
+	 * @param  WC_Product  $product                 Product object to set.
+	 * @param  boolean     $is_forced_subscription  Value.
+	 */
+	public static function set_forced_subscription( $product, $is_forced_subscription ) {
+		self::set_product_property( $product, 'has_forced_subscription', $is_forced_subscription ? 'yes' : 'no' );
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Helpers
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Property getter (compatibility wrapper).
+	 *
+	 * @param  WC_Product  $product   Product object.
+	 * @param  string      $property  Property name.
+	 * @return mixed
+	 */
+	private static function get_product_property( $product, $property ) {
+
+		if ( WCS_ATT_Core_Compatibility::is_wc_version_gte_2_7() ) {
+			$value = $product->get_meta( '_' . $property, true );
+		} else {
+			$value = isset( $product->$property ) ? $product->$property : '';
+		}
+
+		return $value;
+	}
+
+	/**
+	 * Property setter (compatibility wrapper).
+	 *
+	 * @param  WC_Product  $product  Product object.
+	 * @param  string      $name     Property name.
+	 * @param  string      $value    Property value.
+	 * @return mixed
+	 */
+	private static function set_product_property( $product, $name, $value ) {
+
+		if ( WCS_ATT_Core_Compatibility::is_wc_version_gte_2_7() ) {
+			$product->add_meta_data( '_' . $name, $value, true );
+		} else {
+			$product->$name = $value;
+		}
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Filters
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Hooks onto 'woocommerce_is_subscription' to trick Subs into thinking it is dealing with a subscription.
+	 *
+	 * @param  boolean     $is
+	 * @param  int         $product_id
+	 * @param  WC_Product  $product
+	 * @return boolean
+	 */
+	public static function filter_is_subscription( $is, $product_id, $product ) {
+
+		if ( ! $product ) {
+			return $is;
+		}
+
+		if ( self::is_subscription( $product ) ) {
+			$is = true;
+		}
+
+		return $is;
+	}
+
+	/**
+	 * Delete object meta in use by the application layer.
+	 *
+	 * @param  WC_Product  $product
+	 */
+	public static function delete_reserved_meta( $product ) {
+		$reserved_meta_keys = array( 'has_forced_subscription', 'subscription_schemes', 'active_subscription_scheme_key', 'default_subscription_scheme_key' );
+		foreach ( $reserved_meta_keys as $reserved_meta_key ) {
+			$product->delete_meta_data( '_' . $reserved_meta_key );
+		}
```

Comment:
Never done this before. It's the first time I actually need to store "runtime" data on product objects. Using the meta data array to temporarily store "stuff" related to the object is something I have seen in core, but I am having troubling finding where it was. Probably order items? Of course if core devs do something, that doesn't necessarily make it a good example to follow 😄 

Realistically, we have no other way to store "runtime" data on product objects. What we have done here looks safe -- might have been better to use a single meta key to store all data. With that said, I agree that using CRUD to do it is probably a bad idea.

> I don't have a great alternative to suggest, but off the top of my head, it might make sense to store the values in `WCS_ATT_Product` instead.

This would definitely _work_. We could map data to objects using object instance IDs. It still feels more natural to me that object-specific data should be stored _on_ the object. I am really torn here, @thenbrent .